### ISSUE TITLE
Integration with scipy and scikit-learn

### ISSUE THREAD

**Issue by Mec-iS** [DerwenAI/kglab#261]
One of the integration we are going to work on is the one with `scikit-learn`.

This conversation is to collect requirements and features to implement calling `scikit-learn` using `kglab` abstraction layer.

My point of view after taking a look to the API provided by popular data science libraries, these are the interesting `scikit-learn` and `scipy` functionalities that we could start with:
1. Allow converting kglab's `KnowledgeGraph` data structures to observations matrix (to be defined), adjacency matrix and [condensed distance matrix](https://docs.scipy.org/doc/scipy/reference/generated/scipy.spatial.distance.pdist.html) as defined by `scipy`. This will allow building up further flows (or "pipelines", chains of function calls) that the users can assemble to go from a `KnowledgeGraph` representation to a graph algebra representations. This is critical as we need to pick first principles or to provide different alternatives according to the type of graph or the different tasks the users may want to accomplish.
2. After 1, let's start with an example flow in kglab for [SciPy's Hierachical Clustering](https://github.com/scipy/scipy/blob/main/scipy/cluster/hierarchy.py). It would be nice to have a flow that allow simple clustering. This implies providing switches to:
    1. [Linkage procedures](https://github.com/scipy/scipy/blob/main/scipy/cluster/hierarchy.py#L837)
    2. Tree building like [sklearn.cluster.ward_tree](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.ward_tree.html?highlight=graph)

Other possible examples:
* [sklearn.neighbors.kneighbors_graph](https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.kneighbors_graph.html?highlight=graph)
* [Nearest Neighbors](https://scikit-learn.org/stable/modules/neighbors.html?highlight=graph)
* [Agglomerative clustering with and without structure](https://scikit-learn.org/stable/auto_examples/cluster/plot_agglomerative_clustering.html?highlight=graph)
* [Hierarchical clustering: structured vs unstructured ward](https://scikit-learn.org/stable/auto_examples/cluster/plot_ward_structured_vs_unstructured.html?highlight=graph)

These are now in unordered fashion, will take some time to figure out which principles to import from `scikit-learn` and `scipy` so to build up proper user flows from knowledge graph as represented in RDF/kglab and graph algebra representations.

Please provide feedback and suggestions. I will create a Github project around this effort.

cc: @tomaarsen @SultanOrazbayev

**Comment by ceteri** [comment]
Wonderful! This is super helpful.
The nearest neighbor parts would have some immediate use cases.

BTW, there's already the `SubgraphMatrix` class in `subg.py` which handles the transform/inverse_transform from an RDF graph to:

- pandas.DataFrame
- iGraph (adjacency matrix, slightly odd/tangled format)
- NetworkX (adjacency matrix, as an edge list)
- cuGraph (adjacency matrix, as an edge list for cuDF)

**Comment by Mec-iS** [comment]
we probably want some methods that returns `numpy.array`, I will reuse what it is already there for sure.

**Comment by Mec-iS** [comment]
@SultanOrazbayev mentioned the ipmortance of having a descriptive summary of general metrics about a graph, something like `pandas.describe()`. These are the metrics that could be useful in an hypothetical `SubgraphMatrix.describe()`:
* number of nodes, number of edges
* [density](https://networkx.org/documentation/stable/reference/generated/networkx.classes.function.density.html)
* [triangles](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.cluster.triangles.html#networkx.algorithms.cluster.triangles)
* [reciprocity](https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.reciprocity.reciprocity.html#networkx.algorithms.reciprocity.reciprocity)

**Comment by tomaarsen** [comment]
Agreed, sometimes it's hard to actually understand what you're using.
